app/src/org/commcare/views/ViewUtil.java

```diff
@@ -123,4 +85,99 @@ public static int getColorDrawableColor(ColorDrawable drawable) {
             return drawable.getColor();
         }
     }
+
+
+    /**
+     * Determine width of each child view, based on mHints, the suite's size hints.
+     * mHints contains a width hint for each child view, each one of
+     * - A string like "50%", requesting the field take up 50% of the row
+     * - A string like "200", requesting the field take up 200 pixels
+     * - Null, not specifying a width for the field
+     * This function will parcel out requested widths and divide remaining space among unspecified columns.
+     *
+     * @param fullSize Width, in pixels, of the containing row.
+     * @return Array of integers, each corresponding to a child view,
+     * representing the desired width, in pixels, of that view.
+     */
+    public static int[] calculateColumnWidths(ArrayList<String> hints, int fullSize) {
+        // Convert any percentages to pixels. Percentage columns are treated
+        // as percentage of the entire screen width.
+        int[] widths = new int[hints.size()];
+        parseWidths(hints, widths, fullSize);
+
+        Pair<Integer, Integer> constraints = buildConstraints(widths);
+        int claimedSpace = constraints.first;
+        int indeterminateColumns = constraints.second;
+
+        if (claimedSpaceNotTight(fullSize, claimedSpace, indeterminateColumns)) {
+            // Either more space has been claimed than the screen has room for,
+            // or the full width isn't spoken for and there are no indeterminate columns
+            readjustWidths(widths, fullSize, claimedSpace, indeterminateColumns);
+        } else if (indeterminateColumns > 0) {
+            divideIndeterminateSpace(widths, fullSize, claimedSpace, indeterminateColumns);
+        }
+
+        return widths;
+    }
+
+    private static void parseWidths(ArrayList<String> hints, int[] widths, int fullSize) {
+        int hintIndex = 0;
+        for (String hint : hints) {
+            if (hint == null) {
+                widths[hintIndex] = -1;
+            } else if (hint.contains("%")) {
+                String percentString = hint.substring(0, hint.indexOf("%"));
+                widths[hintIndex] = fullSize * Integer.parseInt(percentString) / 100;
+            } else {
+                widths[hintIndex] = Integer.parseInt(hint);
+            }
+            hintIndex++;
+        }
+    }
+
+    private static Pair<Integer, Integer> buildConstraints(int[] widths) {
+        int claimedSpace = 0;
+        int indeterminateColumns = 0;
+        for (int width : widths) {
+            if (width != -1) {
+                claimedSpace += width;
+            } else {
+                indeterminateColumns++;
+            }
+        }
+        return new Pair<>(claimedSpace, indeterminateColumns);
+    }
+
+    private static boolean claimedSpaceNotTight(int fullSize, int claimedSpace,
```

Comment:
The method name here confused me until I thought about it a bit. What about just simplifying to 'widthReadjustmentNeeded', since the comment you have there explains the specifics anyway?